llvm/lib/Transforms/Scalar/Reassociate.cpp

```diff
@@ -539,6 +539,16 @@ static bool LinearizeExprTree(Instruction *I,
     Ops.push_back(std::make_pair(V, Weight));
     if (Opcode == Instruction::Add && Flags.AllKnownNonNegative && Flags.HasNSW)
       Flags.AllKnownNonNegative &= isKnownNonNegative(V, SimplifyQuery(DL));
+    else if (Opcode == Instruction::Mul) {
+      // To preserve NUW we need all inputs non-zero.
+      // To preserve NSW we need all inputs strictly positive.
```

Comment:
Similar to AllKnownNonNegative- its possible for AllKnownNonZero to be `true` in a case where an operand is 0, but `HasNUW` and `HasNSW` are false. Perhaps update the comment on line 71?

nbd, upto you.